Receive GitHub auth tokens with a URI handler rather than a WebSocket

queerviolet opened the pull request:
The good news: We'll no longer be opening a websocket to the auth server. This entirely removes the possibility of an IP mismatch. The code is a lot simpler as well, which is a nice bonus.

The bad news: Both the browser and VS Code will ask if you want to open the custom URI scheme. On top of that, the auth server asks you if you want to allow the extension to access GitHub, and then GitHub asks you if VS Code in general can access GitHub. Nobody asks you if GitHub can access GitHub, but it's only a matter of time.

Fixes #533, #462, #541 (probably! We now trigger the browser immediately, so the auth flow should have a fighting chance to get underway, though there's always the possibility that we'll quickly run aground on a different reef).

/cc: @sguthals @shana @RMacfarlane

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vscode-pull-request-github?pullRequest=547) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vscode-pull-request-github?pullRequest=547) before we can accept your contribution.<br/><br/>:x: queerviolet [sign now](https://cla.opensource.microsoft.com/Microsoft/vscode-pull-request-github?pullRequest=547)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vscode-pull-request-github?pullRequest=547) it.</sub>

queerviolet commented:
<small>oh god</small>

<img width="245" alt="screen shot 2018-10-03 at 11 49 11 pm" src="https://user-images.githubusercontent.com/7763406/46451912-f3762d00-c766-11e8-9bee-3c0fb6834356.png">

queerviolet commented:
Okay, this is ready for review. I've pushed the update to the auth server already, so the new flow should work against the production auth server.

rebornix commented on src/authentication/githubServer.ts:
```diff
@@ -237,27 +119,24 @@ export class GitHubServer {
+			const subscription = uriHandler.event(async uri => {
+				if (uri.path !== CALLBACK_PATH) return
+				const rsp = await axios.get(`${AUTH_RELAY_SERVER}/verify?${uri.query}`)
+				if (rsp.status !== 200) return
```
same as below, we may want to dispose subscription properly.

queerviolet replied:
how embarassing! let us create a rig that ensure this never happens to anyone ever again.

rebornix commented on src/common/uri.ts:
```diff
@@ -101,3 +101,11 @@ export function toPRUri(uri: Uri, pullRequestModel: IPullRequestModel, commit: s
+
+class UriEventHandler extends EventEmitter<Uri> implements UriHandler {
+	public handleUri(uri: Uri) {
+		this.fire(uri)
```
👍

queerviolet replied:
😺

rebornix commented on src/authentication/githubServer.ts:
```diff
@@ -237,27 +119,24 @@ export class GitHubServer {
+		return new Promise<IHostConfiguration>((resolve, reject) => {
+			const subscription = uriHandler.event(async uri => {
+				if (uri.path !== CALLBACK_PATH) return
+				const rsp = await axios.get(`${AUTH_RELAY_SERVER}/verify?${uri.query}`)
```
`axios.get` can throw exceptions, in which case we want to reject the promise as well.

queerviolet replied:
Done (in the adapter logic).

queerviolet commented on src/authentication/githubServer.ts:
```diff
@@ -237,27 +119,24 @@ export class GitHubServer {
+					return reject(new ResponseExpired)
+				resolve({host, username: 'oauth', token})
+				subscription.dispose()
+			})
```
Oops, there is, my lint was off. Fixed em!

queerviolet commented on src/authentication/githubServer.ts:
```diff
@@ -237,27 +119,24 @@ export class GitHubServer {
+				if (rsp.status !== 200) return
+				const {ts, access_token: token} = rsp.data.token
+				if (Date.now() - ts > MAX_TOKEN_RESPONSE_AGE)
+					return reject(new ResponseExpired)
```
fixed it.

queerviolet commented:
Fixed up the subscription handling logic by making a thing that handles the event emitter -> promise conversion. Also fixed up the lint errors.

PTAL 😄

RMacfarlane commented on package.json:
```diff
@@ -297,12 +297,13 @@
   "dependencies": {
     "@octokit/rest": "^15.9.5",
-    "node-emoji": "^1.8.1",
+    "axios": "^0.18.0",
```
Let's also update the `ThirdPartyNotices.txt` with this

rebornix replied:
I updated it.

rebornix commented on src/common/utils.ts:
```diff
@@ -127,4 +127,59 @@ export function formatError(e: any): string {
+			return result;
+		},
+		error => {
+			subscription.dispose();
```
Can we have a wrapper of EventEmitter instead of patching it? EventEmitter is shared in the extension host so we may want to keep it untouched.

queerviolet replied:
I thought that might be controversial 😉 

Done!

queerviolet commented:
Okay, done. PTAL.
